File: java/simple-linked-list/src/main/java/SimpleLinkedList.java
import java.lang.reflect.Array;
import java.util.NoSuchElementException;

class SimpleLinkedList<T> {
    private int length;
    private Node<T> head;

    SimpleLinkedList() {
        this.length = 0;
        this.head = null;
    }

    SimpleLinkedList(T[] values) {
        this.length = 0;

        for (T value : values) {
            this.push(value);
        }
    }

    void push(T value) {
        Node<T> node = new Node(value, this.head);
        this.head = node;
        this.length++;
    }

    T pop() {

        if (this.length == 0) {
            throw new NoSuchElementException();
        }

        T data = this.head.value;
        this.head = this.head.next;
        this.length--;
        return data;
    }

    void reverse() {
        Node<T> p = this.head;
        this.head = null;

        while(p != null) {
            this.push(p.value);
            p = p.next;
        }

    }

    T[] asArray(Class<T> clazz) {
        T[] array = (T[]) Array.newInstance(clazz, this.length);
        Node<T> p = this.head;

        for (int i = 0; i < this.length; i++) {
            array[i] = p.value;
            p = p.next;
        }

        return array;
    }

    int size() {
        return this.length;
    }

    private class Node<T> {

        T value;
        Node<T> next;

        Node (T value, Node<T> next) {
            this.value = value;
            this.next = next;
        }

    }
}

File: python/simple-linked-list/simple_linked_list.py
class Node:
    def __init__(self, value):
        self.data = value
        self.next_node = None

    def value(self):
        return self.data

    def next(self):
        return self.next_node

class LinkedList:
    def __init__(self, values=[]):
        self.count = 0
        self.head_node = None
        self.pointer = None
       
        for value in values:
            self.push(value)

    def __len__(self):
        return self.count

    def __iter__(self):
        return self

    def __next__(self):
        if self.pointer is None:
            raise StopIteration

        node = self.pointer
        self.pointer = self.pointer.next()
        return node.value()

    def head(self):
        if self.count == 0:
            raise EmptyListException("The list is empty.")

        return self.head_node

    def push(self, value):
        node = Node(value)
        node.next_node = self.head_node
        self.head_node = node
        self.pointer = node
        self.count += 1

    def pop(self):
        if self.count == 0:
            raise EmptyListException("The list is empty.")

        data = self.head_node.value()
        self.head_node = self.head_node.next()
        self.pointer = self.pointer.next()
        self.count -= 1
        return data

    def reversed(self):
        return LinkedList(self)

class EmptyListException(Exception):
    def __init__(self, message):
        super().__init__(message)
